maven-model-builder/src/main/java/org/apache/maven/model/building/AbstractModelSourceTransformer.java

```diff
@@ -0,0 +1,196 @@
+        throws TransformerConfigurationException, SAXException, ParserConfigurationException;
+
+    protected OutputStream filterOutputStream( OutputStream outputStream, Path pomFile )
+    {
+        return outputStream;
+    }
+    
+    protected TransformerHandler getTransformerHandler( Path pomFile )
+        throws IOException, org.apache.maven.model.building.TransformerException
+    {
+        return null;
+    }
+
+    @Override
+    public final InputStream transform( Path pomFile, TransformerContext context )
+        throws IOException, org.apache.maven.model.building.TransformerException
+    {
+        final TransformerHandler transformerHandler = getTransformerHandler( pomFile );
+
+        final AbstractSAXFilter filter;
+        try
+        {
+            filter = getSAXFilter( pomFile, context );
+            filter.setLexicalHandler( transformerHandler );
+        }
+        catch ( TransformerConfigurationException | SAXException | ParserConfigurationException e )
+        {
+            throw new org.apache.maven.model.building.TransformerException( e );
+        }
+        
+        final SAXSource transformSource =
+            new SAXSource( filter,
+                           new org.xml.sax.InputSource( new FileInputStream( pomFile.toFile() ) ) );
+
+        final PipedOutputStream pout = new PipedOutputStream();
+        final PipedInputStream pipedInputStream = new PipedInputStream( pout );
+        
+        OutputStream out = filterOutputStream( pout, pomFile );
+
+        final javax.xml.transform.Result result; 
+        if ( transformerHandler == null )
+        {
+            result = new StreamResult( out );
+        }
+        else
+        {
+            result = new SAXResult( transformerHandler );
+            ( (SAXResult) result ).setLexicalHandler( new CommentRenormalizer( transformerHandler ) );
+            transformerHandler.setResult( new StreamResult( out ) );
+        }
+
+        IOExceptionHandler eh = new IOExceptionHandler();
+
+        Thread transformThread = new Thread( () -> 
+        {
+            try ( PipedOutputStream pos = pout )
+            {
+                transformerFactory.newTransformer().transform( transformSource, result );
+            }
+            catch ( TransformerException | IOException e )
+            {
+                eh.uncaughtException( Thread.currentThread(), e );
+            }
+        }, "TransformThread-" + TRANSFORM_THREAD_COUNTER.incrementAndGet() );
+        transformThread.setUncaughtExceptionHandler( eh );
+        transformThread.setDaemon( true );
+        transformThread.start();
+
+        return new ThreadAwareInputStream( pipedInputStream, eh );
+    }
+
+    private static class IOExceptionHandler
+        implements Thread.UncaughtExceptionHandler, AutoCloseable
+    {
+        private volatile Throwable cause;
+
+        @Override
+        public void uncaughtException( Thread t, Throwable e )
+        {
+            try
```

Comment:
Is throwing and catching more efficient than an `instanceof` check?

Reply:
Interesting observation. I borrowed this code, haven't noticed this.